### PR TITLE
Allow Users to see total donations for current year on the Dashboard page

### PR DESCRIPTION
**Pull request by alahab2**
This change creates a Dashboard page which can be accessed by clicking on "HCSA Admin" link in the `navbar`. It also lets Users see the total amount donated for the current year.

See the screenshots below:

screencapture-localhost-3000-donations-1477536828582

---

**Before submitting, check that:**

 - [x] You have added (passing) tests for your code.
 - [x] You have written good* commit messages.
 - [x] You have squashed relevant commits together.
 - [x] You have ensured that RuboCop is passing.
 - [x] Your PR relates to one subject, with a clear title and
       description using complete, grammatically correct sentences.

---

If your change contains views, ensure that:

 - [x] You have included a screenshot of the new view.

---

*If you do not know what makes a good commit message, or why it
is important, please refer to these resources: [1][1], [2][2], [3][3].

[1]: https://robots.thoughtbot.com/5-useful-tips-for-a-better-commit-message
[2]: http://chris.beams.io/posts/git-commit/
[3]: https://github.com/blog/1943-how-to-write-the-perfect-pull-request

- Add link on navbar to dashboard
- Route dashboard to donations/index
- Create donations_for_current_year for dashboard

### CODE REVIEW

File: app/models/donation.rb [hunk 1]
```diff
@@ -9,4 +9,8 @@ class Donation < ActiveRecord::Base
   validates :amount, presence: true, numericality: { greater_than: 0 }
 
   scope :search, -> (keyword) { joins(:donor).where("name ILIKE ?", "%#{keyword}%") }
+
+  def self.donations_for_current_year
```
Comment (Drenmi) [inline app/models/donation.rb]: Let's do two things:

1. Change this into a scope, like the `.search` method.
2. Make it generalized so we can pass a year to it. 😀

File: app/models/donation.rb [hunk 2]
```diff
@@ -9,4 +9,8 @@ class Donation < ActiveRecord::Base
...
+    Donation.where("extract(year  from created_at) = ?", Time.zone.today.year).sum(:amount).round
```
Comment (Drenmi) [inline app/models/donation.rb]: 1. We're already in the `Donation` class, so you don't need to qualify it here.
2. As a convention, use uppercase for SQL keywords: `EXTRACT(YEAR FROM created_at)`
3. `Time.zone.today.year` -> `Date.current.year`

File: app/views/donations/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<div class="card card-inverse card-info text-xs-center">
+  <div class="card-block">
+    <h1> Total Donations for <%= Date.today.year %>: </h1>
```
Comment (Drenmi) [inline app/views/donations/index.html.erb]: `Date.today.year` -> `Date.current.year`

File: spec/factories/donors.rb [hunk 4]
```diff
@@ -16,5 +16,21 @@
         create_list(:donation_with_cause, 2, donor: donor)
       end
     end
+
+    factory :smithie_black do
```
Comment (Drenmi) [inline spec/factories/donors.rb]: Sorry. This is a bit too granular for the factory. We should just pass this as an argument in the test setup. 😀

File: spec/models/donation_spec.rb [hunk 5]
```diff
@@ -9,16 +9,27 @@
   it { is_expected.to validate_numericality_of(:amount).is_greater_than(0) }
 
   describe ".search" do
-    let(:smithie_black) { create(:donor, name: "Smithie Black") }
-    let(:bob_smith_white) { create(:donor, name: "Bob Smith-White") }
-    let(:john_smith) { create(:donor, name: "John Smith") }
-    let(:carl_grey) { create(:donor, name: "Carl Grey") }
+    before do
```
Comment (Drenmi) [inline spec/models/donation_spec.rb]: I would setup this test like so:

```
before do
  create_list(:donor, 2)
end

let(:match_beginning) { create(:donation, create(:donor, "Smithie Black")) }
let(:match_middle) { create(:donation, create(:donor, "Bob Smith White")) }
let(:match_end) { create(:donation, create(:donor, "John Smith")) }

it { expect(described_class.search("smith").to contain_exactly(match_beginning, match_middle, match_end) }
```

That way it is clear which records are meant to match and not, and we can avoid the instance variable confusion. 😀

Comment (alahab2) [inline spec/models/donation_spec.rb]: Then we are facing the same problem with lets that they are created lazily (the one we had last night).
How can we do it without using lets like in .search?

Comment (Drenmi) [inline spec/models/donation_spec.rb]: Should work, because `match_beginning`, `match_middle`, and `match_end` are all referenced in the assertion.

Comment (alahab2) [inline spec/models/donation_spec.rb]: The test passes, but if I remove, for instance, "match_end" from the assertion it still passes.

Comment (alahab2) [inline spec/models/donation_spec.rb]: If I remove the unnecessary factories from `donor`, I can rewrite the test like this:
```
before do
      @match_behinning = create(:donation, donor: create(:donor, name: "Smithie Black"))
      @match_middle = create(:donation, donor: create(:donor, name: "Bob Smith White"))
      @match_end = create(:donation, donor: create(:donor, name: "John Smith"))
      @no_match = create(:donation, donor: create(:donor, name: "Carl Grey"))
    end

    it { expect(described_class.search("smith")).to match_array [@match_behinning, @match_middle, @match_end] } 
```

It is currently passing, and failing if `@no_match` is added to the assertion or `@match_end` is removed unlike the test with `let`s.

File: app/models/donation.rb [hunk 6]
```diff
@@ -9,4 +9,6 @@ class Donation < ActiveRecord::Base
   validates :amount, presence: true, numericality: { greater_than: 0 }
 
   scope :search, -> (keyword) { joins(:donor).where("name ILIKE ?", "%#{keyword}%") }
+
+  scope :donations_per_year, -> (year) { where("EXTRACT(YEAR FROM created_at) = ?", year).sum(:amount).round }
```
Comment (Drenmi) [inline app/models/donation.rb]: Scopes must guarantee to return a relation, so `.sum(:amount).round` needs to live outside of the scope. 😀 Other than that, I'm ready to merge this PR.

Comment (alahab2) [inline app/models/donation.rb]: Does it mean `.sum(:amount).round` should be applied in the view? :thinking:

Comment (Drenmi) [inline app/models/donation.rb]: For now, yes. We might move it to another Ruby object soon. 😀

Comment (alahab2) [inline app/models/donation.rb]: Got it, thanks!

**Pull request merged**